### DISCUSSION TITLE
Too little deformation using DASimpleFoam！

### DISCUSSION THREAD

**Discussion by raisePiginYourhome** [mdolab/dafoam#496] [General]
I have tried to use slsqp and ipopt to optimize it, the primalRes is about 1e-7, the adjoint equation convergence speed of first two iterations could quickly reduce to 1e-6, after that the adjoint equation convergence speed seems too slow:

**Main iteration 0 KSP Residual norm 2.092828485564e-02 5467 s
Main iteration 100 KSP Residual norm 2.092828385109e-02 5585 s
Main iteration 200 KSP Residual norm 2.092806454713e-02 5699 s
Main iteration 300 KSP Residual norm 2.092339861266e-02 5819 s
Main iteration 400 KSP Residual norm 2.090631576810e-02 5946 s
Main iteration 500 KSP Residual norm 2.086681422735e-02 6079 s
Main iteration 600 KSP Residual norm 2.077140236762e-02 6218 s
Main iteration 700 KSP Residual norm 1.949711729518e-02 6362 s
Main iteration 800 KSP Residual norm 1.402296346168e-02 6513 s
Main iteration 900 KSP Residual norm 6.944949875832e-03 6670 s
Main iteration 1000 KSP Residual norm 2.914235117181e-03 6832 s
Main iteration 1100 KSP Residual norm 2.351207525905e-03 6985 s
Main iteration 1200 KSP Residual norm 2.318589501753e-03 7125 s
Main iteration 1300 KSP Residual norm 2.287292387156e-03 7541 s
Main iteration 1400 KSP Residual norm 2.257229369995e-03 10310 s
Main iteration 1500 KSP Residual norm 2.228321421348e-03 16343 s**

I have try the ways by FAQ, so I want to ask is there any other solutions or measures to improve the optimization process.

**Comment by friedenhe** [comment]
Did the flow converge well at the 3rd iteration? Please also give more information, like what case you are running, what is the mesh size, what is the yPlus, and what adjEqnOptions and adjStateOrdering you are using.

**Reply by raisePiginYourhome** [reply]
mesh size: 50thousand
yPlus: min: 0.12 max: 1.13 mean: 0.56
adjEqnOptions: "gmresRelTol": 1.0e-6, "pcFillLevel": 2, "jacMatReOrdering": "natural", "gmresMaxIters": 2000, "gmresRestart": 2000
adjStateOrdering: "cell"

**Comment by friedenhe** [comment]
> Did the flow converge well at the 3rd iteration?

**Reply by raisePiginYourhome** [reply]
yes, 
by the way, I am optimizing a axial pump blade, and I set the physical constrain: flow rate, pressure raise
Here is code:
objFunc:
`"objFunc": {
        "MFR": {
            "part1": {
                "type": "massFlowRate",
                "source": "patchToFace",
                "patches": ["outlet"],
                "scale": 1.0,
                "addToAdjoint": True,
            }
        },
        "PL": {
            "part1": {
                "type": "totalPressure",
                "source": "patchToFace",
                "patches": ["inlet"],
                "scale": -1.0,
                "addToAdjoint": False,
            },
            "part2": {
                "type": "totalPressure",
                "source": "patchToFace",
                "patches": ["outlet"],
                "scale": 1.0,
                "addToAdjoint": False,
            }
        },
        "CMZ": {
            "part1": {
                "type": "moment",
                "source": "patchToFace",
                "patches": ["blade"],
                "axis": [0.0, 0.0, 1.0],
                "center": [0.0, 0.0, 0.0],
                "scale": 1.0*1000,
                "addToAdjoint": True,
            }
        },
    },`
physical constraints:
`optProb.addCon("PL", lower=51, upper=51, scale=1)`

The turbulenceModel is SpalartAllmaras, and the primal flow converges well, the ResTol is 1e-6.
But when evaluating objective function sensitivities, the KSP Residual of three linear equations are:
`Main iteration 0 KSP Residual norm 0.000000000000e+00 2218 s
Main iteration 0 KSP Residual norm 0.000000000000e+00 2218 s `
If I removed the pressure raise constrain, the moment objectFunc could work well:
`    Adding all faces of patches 1(blade) ...
    Found matching patch blade with 2816 faces.
Solving Linear Equation... 2308 s
Main iteration 0 KSP Residual norm 5.500082557991e+02 2308 s
Main iteration 100 KSP Residual norm 1.667436391357e-02 2367 s
Main iteration 200 KSP Residual norm 6.720682445976e-04 2430 s
Main iteration 222 KSP Residual norm 5.344737313302e-04 2444 s 
Total iterations 222
Solving Linear Equation... Completed! 2444 s
Residual tolerance satisfied, solution finished!`

So is there anyway or solution to fix this situation! Looking forward to your response.

**Comment by friedenhe** [comment]
Not sure what caused this... Can you share the log file?

**Reply by lnhang0618** [reply]
When the meshWaveFrozen method is utilized for wallDist, it appears that the adjoint equations converge slowly

**Comment by friedenhe** [comment]
@lnhang0618 This is not expected... Did the meshWave method converge well? Did the adjoint converge slowly at the first iteration or later?

**Reply by lnhang0618** [reply]
I'm not sure, I can't reproduce this problem anymore. I think if I meet it next time, I'll consider investigating further, but for now, it doesn't seem necessary to delve deeper into it.
